core/desugarFuns.ml

```diff
@@ -20,66 +25,78 @@ open SugarConstructors.DummyPositions
       f2
     }
 
+  2. Anonymous functions are named and then turned into a series of nested
+     one-argument functions:
+
 
     fun [qs](xs1)...(xsk) {e}
   -->
-    {fun f[qs](xs1)...(xsk) {e};
-     f}
+    {fun f[qs](xs1) {
+       fun f1[](xs2) {
+       ...
+         fun fk[](xsk) {
+           e
+         }; fk
+       ...
+       }; f1
+    }; f}
 
 
+  3. Operator sections are turned into named functions:
+
     (.l)
   -->
-    fun (r) {r.l}
+    { fun f(x:r) {x.l}; f }
+
 *)
 
 
 (* unwrap a curried function definition as
    a collection of nested functions
 *)
-let unwrap_def ({node=f, ft; _}, linearity, (tyvars, lam), location, t) =
+let unwrap_def ({node=f, ft; _}, linearity, lam, location, t) =
   let ft = val_of ft in
   let rt = TypeUtils.return_type ft in
   let lam =
     let rec make_lam t : funlit -> funlit =
       function
-        | ([_ps], _body) as lam -> lam
-        | (ps::pss, body) ->
+        | (_, [_ps], _body) as lam -> lam
+        | ([], ps::pss, body) ->
```

Comment:
It looks like there are many assertion failures on line 71 due to this pattern match being non-exhaustive.  Is there a particular reason we can assume that the tyvar list is empty if the first pattern doesn't match?  What would be wrong with binding it to a variable and using that variable again on line 67 below?

Reply:
OK, so I tried this (binding `tyvars` instead of [], and passing `tyvars` through on line 67) and while more tests pass, there are still a bunch of ArityMismatch errors.  So not clear this is the right thing to do either, even by randomly-hacking-and-seeing-if-tests-pass standards.

Reply:
Yes, good point. This is basically a work in progress towards a proper solution and I haven't really bothered to test before pushing :-)

Reply:
> there are still a bunch of ArityMismatch errors

Yes, these are expected at this point.